### DISCUSSION TITLE
How to work around a significant DataFetchingFieldSelectionSet performance issue with many types

### DISCUSSION THREAD

**Discussion by aleksandarsusnjar** [graphql-java/graphql-java#2309] [Q&A] [answered]
Hi! I am playing with GraphQL-Java and have had really good progress really fast. Perhaps too fast.

I've landed on an issue where `DataFetchingFieldSelectionSetImpl.getImmediateFields()` is taking 20 seconds to run. I cannot work with that and need help working around the issue. Inspection showed that the result from it had 3.5 million entries in my case for a GraphQL query that looks roughly like this:

```
{
  things {
    id
    parent {
      id
      name
      parent {
        id
        name
        parent {
          id
          name
          parent {
            id
            name
          }
        }
      }
    }
  }
}
```

Yes, this is "deep" but quite a simple query otherwise. My gut feeling about what sets this off is that we have over a hundred various types (interfaces and objects) that any parent can be be of and I haven't found a way to simply ask for "give me what is stated, don't try to resolve this to per-type fields" which causes huge repetition - each field is listed many times, one per type it exists in, for all possible types, at each level.

I cannot change the data model - this is a complex, large application and the types in it exist for good reasons. Dividing this time into smaller chunks wouldn't be helpful either, as it would still be spent, just differently. How can I get what is requested by the client at "face value" and no processing/resolving at all?

**Comment by andimarek** [comment]
Can you please share you schema? If you don't want to share it publicly please have a look at https://github.com/graphql-java/graphql-anonymizer. Thanks

**Reply by aleksandarsusnjar** [reply]
Will try the anonymizer.

**Comment by aleksandarsusnjar** [comment]
Here.

[anonymized.zip](https://github.com/graphql-java/graphql-java/files/6387231/anonymized.zip)

**Comment by andimarek** [comment]
Can you also please share the query you are executing?

**Comment by aleksandarsusnjar** [comment]
Note that the schema is built programmatically. What I attached is the anonymized version of what IntelliJ IDEA GraphQL plugin obtains from the servlet. The query is at the bottom, I did anonymize it too.

**Reply by andimarek** [reply]
Thanks!

**Comment by aleksandarsusnjar** [comment]
It is at the bottom of the file:
```
query {field1121 {field10 field50 {field10 field49 field50 {field10 field49 field50 {field10 field49 field50 {field10 field49}}}}}}
```

**Comment by aleksandarsusnjar** [comment]
I have some means to generate the schema differently but the type hierarchy is extensive. I presently convert concrete leaf types to GraphQL object types and all others are interfaces.

My code is trying to do what is described here:

https://www.graphql-java.com/blog/deep-dive-merged-fields/

... at large scale. I already have the code that prefetches everything it needs upfront... but is held up / delayed by getImmediateFields().

If it helps, the code is roughly (unrolled from actual):

```java
        final DataFetchingFieldSelectionSet selectionSet = environment.getSelectionSet();
        final List<SelectedField> immediateFields = selectionSet.getImmediateFields();
```

**Comment by aleksandarsusnjar** [comment]
If it can help ... I ran a jProfiler on this (so times are inflated)... 

![Screen Shot 2021-04-27 at 4 43 14 PM](https://user-images.githubusercontent.com/40366235/116310229-b84e0600-a777-11eb-9c33-6f4f4c4af6a7.png)

![Screen Shot 2021-04-27 at 4 44 48 PM](https://user-images.githubusercontent.com/40366235/116310379-ecc1c200-a777-11eb-892a-b10990c80e5e.png)

![Screen Shot 2021-04-27 at 4 47 31 PM](https://user-images.githubusercontent.com/40366235/116310640-47f3b480-a778-11eb-8fbf-417aa2d9e32c.png)

**Comment by aleksandarsusnjar** [comment]
Complete call tree... 
[Call_Tree_Complete.zip](https://github.com/graphql-java/graphql-java/files/6387596/Call_Tree_Complete.zip)

**Reply by aleksandarsusnjar** [reply]
Is there a way I can avoid this "normalization" process? I have a model already and know where things are coming from. This `NormalizedQueryTreeFactory.buildFieldWithChildren(...)` process may be useful to someone (else) but is extremely expensive for me...

Another way to ask the same thing... is there a way to simply get exactly what is stated in the original query - nothing more, nothing less. If things are stated multiple times with different arguments and for different types, sure. But don't try to resolve these to all possible types out there. In fact, these are mapped to a succinct Java class hierarchy where each field is declared exactly once and I already have that model (from which I am generating the GraphQL-Java schema). Heck, I can even validate it if validation is an issue...

I looked at `contains...(...)`, `getFields(...)`, `getFieldsGroupedByResultKey(...)` but all these make a call to the expensive `computeValuesLazily()`...

Should I be looking into using one of the following?

1. `environment.getContext()` - which context is this? This is different than `getLocalContext()`
2. `environment.getDocument()` and iterate over nodes myself? Is there something in between?
3. `environment.getExecutionStepInfo()` - or something like that? But would this have the same issue?

(You may see that I am still learning how to use `graphql-java`. I am learning about GraphQL in parallel. This business without interfaces extending other interfaces and having to declare all supertypes and redeclare all fields from supertypes did surprise me and is possibly causing the explosion of the "selection set")

**Comment by aleksandarsusnjar** [comment]
Note - I am trying to rework the schema a bit so that the collection fields use more specific types. That isn't trivial as many types are navigable from many others (i.e. it is a many-to-many type-to-type relationship). This will actually essentially triple the number of GraphQL schema over what I really have in the system but each field will only yield a smaller subset. That subset may still be significant, but will be lesser than what you can see in this schema. I think that should help a bit but it may not be enough.

I could try to help but I am anxious about doing so as I don't know enough about what is going on in there and this feels like it is important enough to get done "right" having in mind whatever future you have envisioned for this...

**Comment by aleksandarsusnjar** [comment]
Any suggestions here? I did attempt to focus/constrain the relationships but there are multiple kinds of them that would require that and I only did one. How can I get a field selection without having it normalized?

**Comment by aleksandarsusnjar** [comment]
After constraining relationships, now my proof of concept is facing a different issue. I need to be able to specify subtypes to access some properties, i.e.:

```GraphQL
{
    things {
        id
        parent {
            id
            name
            ... on ParentSubtype1 {
                subtypeOnlyField1
                subtypeOnlyField2
            }
        }
    }
}
```

But I don't know how to get  the information as presented. I'd expect/like the equivalent of a collection/map as follows:

**1. Properties requested for the base type - immediate**
```GraphQL
    ... on ParentBaseType { # implied
        id
        name
    }
```

**2. Properties requested for the `ParentSubtype1`**
```GraphQL
    ... on ParentSubtype1 {
        subtypeOnlyField1
        subtypeOnlyField2
    }
```

... and not a "normalized" exploded version that I am able to get (imagine if `... on ___` wasn't polymorphic):
```GraphQL
parent {
     # id
     ... on ParentBaseType { id }
     # ... then repeated for all subypes of ParentBaseType... 
     ... on ParentSubtype1 { id }
     ... on ParentSubtype2 { id }
     #...
     ... on ParentSubtypeN { id }

     # name
     ... on ParentBaseType { name }
     # ... then repeated for all subypes of ParentBaseType...
     ... on ParentSubtype1 { name }
     ... on ParentSubtype2 { name }
     #...
     ... on ParentSubtypeN { name }

     # ... on ParentSubtype1 { subtypeOnlyField1 }
     ... on ParentSubtype1 { subtypeOnlyField1 }
     # ... then repeated for all subypes of ParentSubtype1...
     ... on ParentSubtype3 { subtypeOnlyField1 }
     #...
     ... on ParentSubtypeX { subtypeOnlyField1 }
     
     # ... on ParentSubtype1 { subtypeOnlyField2 }
     ... on ParentSubtype1 { subtypeOnlyField2 }
     # ... then repeated for all subypes of ParentSubtype1...
     ... on ParentSubtype3 { subtypeOnlyField2 }
     #...
     ... on ParentSubtypeX { subtypeOnlyField2 }
}
```

**Reply by aleksandarsusnjar** [reply]
(Please help... even if it is presently impossible, I need to know that I am not chasing something that isn't there...)

**Comment by aleksandarsusnjar** [comment, marked as answer]
Seems related: https://github.com/graphql-java/graphql-java/issues/2185
